cty/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+//! Type aliases for C FFI
+#![cfg_attr(not(feature = "rustc-dep-of-std"), deny(warnings))]
+#![allow(non_camel_case_types)]
+#![cfg_attr(
+    feature = "rustc-dep-of-std",
+    feature(no_core)
+)]
+#![no_std]
+#![cfg_attr(feature = "rustc-dep-of-std", no_core)]
+
+#[macro_use]
+mod macros;
+
+/// Architecture independent
+mod ai {
+    pub type int8_t = i8;
+    pub type int16_t = i16;
+    pub type int32_t = i32;
+    pub type int64_t = i64;
+
+    pub type uint8_t = u8;
+    pub type uint16_t = u16;
+    pub type uint32_t = u32;
+    pub type uint64_t = u64;
+
+    pub type c_schar = i8;
+    pub type c_short = i16;
+    pub type c_longlong = i64;
+
+    pub type c_uchar = u8;
+    pub type c_ushort = u16;
+    pub type c_ulonglong = u64;
+
+    pub type c_float = f32;
+    pub type c_double = f64;
+
+    pub type intmax_t = i64;
+    pub type uintmax_t = u64;
+
+    pub type size_t = usize;
+    pub type ptrdiff_t = isize;
+    pub type intptr_t = isize;
+    pub type uintptr_t = usize;
+
+    cfg_if! {
+        if #[cfg(libc_core_cvoid)] {
+            pub use core::ffi::c_void;
+        } else {
+            // Use repr(u8) as LLVM expects `void*` to be the same as `i8*` to help
+            // enable more optimization opportunities around it recognizing things
+            // like malloc/free.
+            #[repr(u8)]
+            #[allow(missing_copy_implementations)]
+            #[allow(missing_debug_implementations)]
+            pub enum c_void {
+                // Two dummy variants so the #[repr] attribute can be used.
+                #[doc(hidden)]
+                __variant1,
+                #[doc(hidden)]
+                __variant2,
+            }
+        }
+    }
+}
+
+/// Architecture dependent
+mod ad {
+    cfg_if! {
+        if #[cfg(any(target_arch = "aarch64",
+                     target_arch = "arm",
+                     target_arch = "asmjs",
+                     target_arch = "wasm32",
+                     target_arch = "wasm64",
+                     target_arch = "powerpc",
+                     target_arch = "powerpc64",
+                     target_arch = "s390x"))] {
+            cfg_if! {
+                if #[cfg(not(any(target_os = "apple", target_os = "ios")))] {
+                    pub type c_char = ::c_uchar;
+
+                    cfg_if! {
+                        if #[cfg(any(
+                            target_arch = "powerpc",
+                            target_arch = "powerpc64",
+                            target_arch = "s390x",
+                        ))] {
+                            // FIXME: ??
+                            pub type wchar_t = i32;
+                        } else {
+                            pub type wchar_t = u32;
```

Comment:
Windows also supports `aarch64` and `arm` where `wchar_t` is still `u16`.

Reply:
I think `wchar_t` is just too OS and target-dependent to try to shoe-horn it here.